Dockerfile

```diff
@@ -1,31 +1,121 @@
+        -DPYTHON_INSTALL_DIR=$(python3 -c 'import sys; print(f"/usr/lib/python{sys.version_info.major}.{sys.version_info.minor}/dist-packages")') && \
+    make -j$(($(nproc)-1)) && \
+    make install && \
+    cp ../config/99-realsense-libusb.rules /etc/udev/rules.d/ && \
+    cp ../config/99-realsense-d4xx-mipi-dfu.rules /etc/udev/rules.d/ && \
+    cd / && \
+    rm -rf librealsense 
+
+ENV LD_LIBRARY_PATH /usr/local/lib:$LD_LIBRARY_PATH
+
+# Docker complains about missing UDEV rules but being root, it is not a problem
+# RUN service udev start && udevadm control --reload-rules && udevadm trigger
+
+# Setup ROS2 environment to compile ROS packages
+RUN pip install \
+        colcon-common-extensions \
+        colcon-mixin 
 
-# setup colcon mixin and metadata
 RUN colcon mixin add default \
-      https://raw.githubusercontent.com/colcon/colcon-mixin-repository/master/index.yaml && \
+        https://raw.githubusercontent.com/colcon/colcon-mixin-repository/master/index.yaml && \
     colcon mixin update && \
     colcon metadata add default \
-      https://raw.githubusercontent.com/colcon/colcon-metadata-repository/master/index.yaml && \
+        https://raw.githubusercontent.com/colcon/colcon-metadata-repository/master/index.yaml && \
     colcon metadata update
 
-# copy workspace
+RUN apt update && apt install python3-rosdep -y 
+
+ENV SKIP_KEYS "libopencv-dev libopencv-contrib-dev libopencv-imgproc-dev python-opencv python3-opencv librealsense2"
+
+# Compile realsense-ros
+RUN source /ros_entrypoint.sh && \
+    mkdir -p /realsense-ros/src && \
+    cd /realsense-ros && \
+    rosinstall_generator --rosdistro ${ROS_DISTRO} \
+        xacro \
+        diagnostic_updater \
+        sensor_msgs_py \
+        launch_pytest \
+        realsense2_camera \
+        realsense2_camera_msgs \
+        realsense2_description \
+        > realsense.rosinstall && \
+    vcs import src < realsense.rosinstall && \
+    rosdep install -i --from-path src --rosdistro $ROS_DISTRO --skip-keys="$SKIP_KEYS" -y && \
+    colcon build --merge-install --cmake-args -DCMAKE_BUILD_TYPE=Release && \
+    sed -i "\$i ros_source_env /realsense-ros/install/local_setup.bash \n" /ros_entrypoint.sh
+
+# Compile image compression plugins (image-transport-plugins)
+RUN source /ros_entrypoint.sh && \
+    mkdir -p /image-transport-plugins/src && \
+    cd /image-transport-plugins && \
+    rosinstall_generator --rosdistro ${ROS_DISTRO} \
+        image_transport_plugins \
+        theora_image_transport \
+        > image-transport-plugins.rosinstall && \
+    vcs import src < image-transport-plugins.rosinstall && \
+    rosdep install -i --from-path src --rosdistro $ROS_DISTRO -y --skip-keys="$SKIP_KEYS" && \
+    colcon build --merge-install --cmake-args -DCMAKE_BUILD_TYPE=Release && \
+    sed -i "\$i ros_source_env /image-transport-plugins/install/local_setup.bash \n" /ros_entrypoint.sh
+
+
+RUN pip install python-can
+
+FROM reseq-base AS reseq
```

Comment:
Why adding a second `FROM` using the just built image? AFAIK this doesn't give you any advantage. Instead you could do a second `FROM` but from `${BASE_IMAGE}` instead, and copy the artifacts that you built on the first stage. I'm not sure this is possible with colcon packages, but it would much improve the final image.

Reply:
The reason for the second image is to allow building only the first part of the Dockerfile (using the target parameter) in order to test the dependencies without having our files in the way which execute on startup or having fails in colcon if the dockerfile is not correct. This is especially useful when building new packages in ROS since the procedure to determine the dependencies that rosdep is unable to resolve needs some trial and error (at least that is what I understood of it). 

I agree that it would be better to do what you suggest but that would require to identify which libraries are used by the realsense only at compile-time and which one are also used at runtime, and it is not that easy to do, so I preferrred to avoid wasting time doing that,

Also it is worth noting that in the future we won't use this dockerfile because the NVIDIA ISAAC tools require a completely different approach to be used so as a temporary solution ir seemed good enough

Reply:
Ok, like this makes much more sense! Thanks for the explanation.

Yeah if the plan is to change the image it's a valid solution. If you end up needing to optimize it I could definitely help!

Reply:
Yes, the idea is to use the [NVIDIA ROS buildfarm](https://nvidia-isaac-ros.github.io/getting_started/isaac_ros_buildfarm_cdn.html) image which allow us to transfer compressed images without CPU saturation using the ISAAC GPU tools that afaik are essentially only availbale there. That image would also have the advantage of having some ROS packages installable with APT. The only problems are that it requires an SSD (this also explains the request to buy it) and re-flashing the Jetson to another version of te JetPack (5.1.2) (which we prefer to postpone to avoid having zero Jetsons working)

Reply:
If I recall correctly you can update the JetPack version without reinstalling, by just thinkering a bit through `apt`.